### PR TITLE
[FEATURE] std::views::join_with

### PR DESCRIPTION
**Pull request by eseiler**
Implemented according to https://eel.is/c++draft/range.join.with

Structure is similar to the zip_view.

### CODE REVIEW

**Review by SGSSGene — APPROVED** [review]

File: include/seqan3/utility/views/join_with.hpp [hunk 1]
```diff
@@ -6,26 +6,525 @@
...
+    constexpr V base() &&
+    {
+        return std::move(base_);
+    }
+
+    constexpr auto begin()
+    {
+        constexpr bool use_const =
+            view_helper::simple_view<V> && std::is_reference_v<InnerRng> && view_helper::simple_view<Pattern>;
+        return iterator<use_const>{*this, std::ranges::begin(base_)};
```
Comment (smehringer) [inline include/seqan3/utility/views/join_with.hpp:112]: why do we enforce const? (not sure what the above `use_const` means semantically)

File: include/seqan3/utility/views/join_with.hpp [hunk 2]
```diff
@@ -6,26 +6,525 @@
...
+
+    using OuterIter = helper_t::OuterIter;
+    using InnerIter = helper_t::InnerIter;
+    using PatternIter = helper_t::PatternIter;
+
+    static constexpr bool ref_is_glvalue = helper_t::ref_is_glvalue;
+
+    friend class join_with_view<V, Pattern>;
+
+    Parent * parent_ = nullptr;
```
Comment (smehringer) [inline include/seqan3/utility/views/join_with.hpp:259]:
```suggestion
    Parent * parent_{nullptr};
```
?

File: include/seqan3/utility/views/join_with.hpp [hunk 3]
```diff
@@ -6,26 +6,525 @@
...
+
+public:
+    OuterIter outer_it_ = OuterIter();
```
Comment (smehringer) [inline include/seqan3/utility/views/join_with.hpp:262]:
```suggestion
    OuterIter outer_it_{};
```
?

File: include/seqan3/utility/views/join_with.hpp [hunk 4]
```diff
@@ -6,26 +6,525 @@
...
+
+    constexpr auto && get_inner(OuterIter const & x)
+    {
+        if constexpr (ref_is_glvalue)
+            return *x;
+        else
+            return *parent_->inner_;
+    }
+
+    constexpr void satisfy()
```
Comment (smehringer) [inline include/seqan3/utility/views/join_with.hpp:295]: can you add a short brief here? it's hard to review by just looking at the code.

File: include/seqan3/utility/views/join_with.hpp [hunk 5]
```diff
@@ -6,26 +6,525 @@
...
+    }
+
+    constexpr iterator & operator++()
+    {
+        std::visit(
+            [](auto & it)
+            {
+                ++it;
+            },
+            inner_it_);
```
Comment (smehringer) [inline include/seqan3/utility/views/join_with.hpp:372]: doesn't this fit on one line

**Review by smehringer — COMMENTED** [review]
Did you test a temporary input, and an input that is non const but triggers the "enforce const" version?

File: include/seqan3/utility/views/join_with.hpp (hunk 5 above)
Comment (eseiler) [reply include/seqan3/utility/views/join_with.hpp:374]: Only if I prevent clang-format from changing it back; should I?

**Review by smehringer — APPROVED** [review]

**Pull request merged**